src/main/java/com/oltpbenchmark/benchmarks/templated/TemplatedBenchmark.java

```diff
@@ -0,0 +1,255 @@
+                }
+                generators.put(proc.getClass(), new TraceTransactionGenerator(list));
+            }
+
+            // Create workers.
+            int numTerminals = workConf.getTerminals();
+            LOG.info(String.format("Creating %d workers for templated benchmark", numTerminals));
+            for (int i = 0; i < numTerminals; i++) {
+                workers.add(new TemplatedWorker(this, i, generators));
+            }
+        } catch (Exception e) {
+            throw new IllegalStateException("Unable to create workers", e);
+        }
+        return workers;
+    }
+
+    @Override
+    protected Loader<TemplatedBenchmark> makeLoaderImpl() {
+        // Loader not currently supported.
+        throw new UnsupportedOperationException();
+    }
+
+    private ClassLoader loadQueryTemplates(String file) {
+        // Instantiate Java compiler.
+        CustomClassLoader ccloader = new CustomClassLoader(this.classLoader);
+        try {
+            // Parse template file.
+            final ICompilerFactory compilerFactory = CompilerFactoryFactory.getDefaultCompilerFactory(
+                    TemplatedBenchmark.class.getClassLoader());
+
+            JAXBContext jc = JAXBContext.newInstance("com.oltpbenchmark.api.templates");
+            SchemaFactory sf = SchemaFactory.newInstance(javax.xml.XMLConstants.W3C_XML_SCHEMA_NS_URI);
+            Schema schema = sf.newSchema(new StreamSource(this.getClass().getResourceAsStream("/templates.xsd")));
+            Unmarshaller unmarshaller = jc.createUnmarshaller();
+            unmarshaller.setSchema(schema);
+
+            StreamSource streamSource = new StreamSource(new FileInputStream(file));
+            JAXBElement<TemplatesType> result = unmarshaller.unmarshal(streamSource, TemplatesType.class);
+            TemplatesType templates = result.getValue();
+
+            for (TemplateType template : templates.getTemplateList()) {
+                ImmutableParsedQueryTemplate.Builder b = ImmutableParsedQueryTemplate.builder();
+                b.name(template.getName());
+                b.query(template.getQuery());
+                b.paramsTypes(template.getTypes().getTypeList());
+                for (ValuesType paramValue : template.getValues()) {
+                    b.addParamsValues(String.join(", ", paramValue.getValueList()));
+                }
+
+                ParsedQueryTemplate qt = b.build();
+                // Create and compile class.
+                final String s = """
+                        package %s ;
+                        public final class %s extends %s {
+                            @Override
+                            public %s getQueryTemplateInfo() {
+                                return ImmutableQueryTemplateInfo.builder()
+                                        .query(new %s(\"%s\"))
+                                        .paramsTypes(new String[] {%s})
+                                        .paramsValues(new String[] {%s})
```

Comment:
For future NULLable handling, I wonder if it might be possible to change this to pass a multi-dimensional array of objects from the XML parser so that the CSV re-parsing portion can be skipped.
That would be a step towards supporting passing NULLs in, but I think the XML schema might also have to be adjusted somehow to account for it - right now the values received are `""` even when using something like `<value />`

Reply:
Right now we're reading the input as a String[], then use the CSV Parser to parse it into an Object[]. The actual type is not set until the query is triggered, using the SQL.Type to set the object via the JDBC driver. There's two pain points: The instantiation into a class which we currently do as String[] to avoid any object-specific handling and then the conversion into the Object[]. The CSV Parser was initially designed to help with parsing different types of objects, if memory serves, but I agree that it's convoluted.

Reply:
Good notes for the future.  Thanks.  Hopefully whoever takes this issue (#366) up later will read these threads again :)